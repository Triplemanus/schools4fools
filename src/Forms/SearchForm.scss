header {
  align-items: center;
  height: 25vh;
  margin:auto;
  width: 80%;
}

// h1 {
//   height: 10%;
//   color: #3498DB;
//   padding: 30px;
//   font-size: 64px;
// }

h1 span {
  color: #E74C3C;
}

#search-form {
  align-items: center;
  display: flex;
  height: 2em;
  margin-left: 2%;
  text-align: center;
}

input, form button {
  font-size: 14px;
  height: 2em;
  width: 20%;
}

input {
  border: 2px solid #A5A5A5;
  padding-left: 10px;
}

#search-form button {
  border: 2px solid #A5A5A5;
  background-color: #3498DB;
  font-size: 14px;
  font-weight: bold;
  color: #FFF;
  height: 2.4em;
}