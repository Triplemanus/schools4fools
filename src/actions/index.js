export const isLoading = isLoading => ({
  type: 'IS_LOADING'
});

export const getAllSchools = (schools) => ({
  type: 'GET_SCHOOLS',
  schools
});

export const hasErrored = errorMsg => ({
  type: 'HAS_ERRORED',
  errorMsg
});

export const addSchools = schools => ({
  type: 'ADD_SCHOOLS',
  schools
});