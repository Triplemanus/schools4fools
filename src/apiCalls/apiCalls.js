export const fetchAllSchools = async (querySchools) => {
  const url = `https://api.schooldigger.com/v1.2/schools?st=${querySchools.state}&level=${querySchools.level}&isMagnet=false&isCharter=false&isVirtual=false&isTitleI=true&isTitleISchoolwide=false&nearLatitude=${querySchools.locLatitude}&nearLongitude=${querySchools.locLongitude}&distanceMiles=${querySchools.maxDistance}&perPage=30&sortBy=distance&appID=d04b0481&appKey=0c80520db8adfa4e8be1ea61724e1f24`;
  //const url = 'https://api.schooldigger.com/v1.2/schools?st=CO&level=High&nearLatitude=39.7506294&nearLongitude=-104.9966207&distanceMiles=20&appID=d04b0481&appKey=0c80520db8adfa4e8be1ea61724e1f24';
  //console.log('querySchools latLong: ', querySchools, querySchools.latLocation, querySchools.longLocation);
  //console.log('fetchAllSchools url is ', url);
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error ('There was an error getting your school data.');
  }
  const schools = await response.json();
  return schools;
}