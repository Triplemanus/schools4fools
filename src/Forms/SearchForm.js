import React from 'react';
import { connect } from 'react-redux';
import { addSchools } from '../actions/index';
import { fetchAllSchools } from '../apiCalls/apiCalls';

export class SearchForm extends React.Component {
  constructor() {
    super();
    this.state = {
      locState: '',
      level: '',
      latLocation: '',
      longLocation: '',
      maxDistance: '',
      errorMsg: ''
    }
  };

  updateForm = e => {
    this.setState({ [e.target.name]: [e.target.value] });
    console.log('updateForm state, locState is:', this.state, this.state.locState);
  };

  clearInputs = () => {
    this.setState({ 
      locState: '',
      level: '',
      latLocation: '',
      longLocation: '',
      maxDistance: '',
      error: ''
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const querySchools = {
      state: this.state.locState,
      level: this.state.level,
      latLocation: parseFloat(this.state.latLocation),
      longLocation:  parseFloat(this.state.longLocation),
      maxDistance: parseInt(this.state.maxDistance)
    };
    console.log('querySchools in submit: ', querySchools);
    this.searchSchools(querySchools);
    this.clearInputs();
  };

  searchSchools = querySchools => {
    console.log('querySchools in searchSchools: ', querySchools);
    fetchAllSchools(querySchools)
    .then(schools => {
      console.log('schools: ', schools);
      this.props.addSchools(schools)})
    .catch(error => this.setState({ error: error }));
  };

  render() {
    const { locState, level, latLocation, longLocation, maxDistance } = this.state;
    console.log('SearchForm state is : ', this.state);
    return (
      <header>
        <h1>Search Form</h1>
        <form id="SearchForm">
          <input
            type="text"
            name="locState"
            value={locState}
            placeholder="State"
            onChange={this.updateForm}        
          />
          <input
            name="level"
            value={level}
            placeholder="Level"
            onChange={this.updateForm}          
          />
          <input
          name="latLocation"
          value={latLocation}
          placeholder="Latitude"
          onChange={this.updateForm}
          />
          <input
          name="longLocation"
          value={longLocation}
          placeholder="Longitude"
          onChange={this.updateForm}
          />
          <input
          name="maxDistance"
          value={maxDistance}
          placeholder="Distance"
          onChange={this.updateForm}
          />
          <button onClick={event => this.handleSubmit(event)}>
            Search Schools
          </button>
        </form>
      </header>
    )
  }
};

export const mapStateToProps = state => ({
  ...state
});

export const mapDispatchToProps = dispatch => ({
  addSchools: schools => dispatch(addSchools(schools))
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchForm);