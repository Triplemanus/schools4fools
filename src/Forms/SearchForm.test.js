import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapStateToProps, mapDispatchToProps } from './SearchForm';
import { fetchAllSchools } from '../apiCalls/apiCalls';

import { addSchools } from '../actions/index'; 

jest.mock('../apiCalls/apiCalls.js');

describe('SearchForm', () => {
  let wrapper, mockSchool, mockAddSchools;

  beforeEach(() => {
    mockSchool = null;
    mockAddSchools = jest.fn();
    wrapper = shallow(<SearchForm 
    school={mockSchool}
    addSchools={mockAddSchools}
    />);

    fetchAllSchools.mockImplementation(() => {
      return Promise.resolve([{
        school_id: "080336006756",
        key: "080336006756",
        school_name: "George Washington",
        phone: '303.399.4532',
        address: '1234 S Monaco St.',
        distance: 5.8,
        low_Grade: '9',
        high_Grade: '12',
        level: 'High',
        is_Charter: false,
        is_Magnet: false,
        is_Private: false
      }])
    })
  });

  it('should match the snapshot with all data passed in correctly', () => {

    expect(wrapper).toMatchSnapshot();
  });
  
  it('should update state on a change in an input', () => {
    
    const mockEvent = { target: { name: 'level', value: 'High' } };
    const mockEventDistance = { target: { name: 'distance', value: 20 } }
    
    wrapper.find('input').at(2).simulate('change', mockEvent);
    wrapper.find('input').at(4).simulate('change', mockEventDistance);
    
    expect(wrapper.state('level')).toEqual(['High']);
    expect(wrapper.state('distance')).toEqual([20]);
  });

  it('should call handleSubmit on a button click', () => {

    const mockEvent = {target: {name: 'locState', value: 'CO'}}

    wrapper.instance().handleSubmit = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should update the state when clearInputs is called', () => {

    const expected = '';

    wrapper.setState({
      locState: 'CO',
      error: 'These are not the schools you are looking for.'
    });

    wrapper.instance().clearInputs();

    expect(wrapper.state('locState')).toEqual(expected);
    expect(wrapper.state('error')).toEqual(expected);
  });

  it('should run searchSchools on a button click', () => {
    const mockEvent = { preventDefault: jest.fn() }

    wrapper.instance().searchSchools = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().searchSchools).toHaveBeenCalled();
  });
});

describe('mapStateToProps', () => {

  it('should return an object with the current school', () => {
    const mockState = {
      error: '',
      schools: [
        {
          school_id: "080336006756",
          key: "080336006756",
          school_name: "George Washington",
          phone: '303.399.4532',
          address: '1234 S Monaco St.',
          distance: 5.8,
          low_Grade: '9',
          high_Grade: '12',
          level: 'High',
          is_Charter: false,
          is_Magnet: false,
          is_Private: false
        }
      ],
    }

    const expectedSchools =  [{
      school_id: "080336006756",
      key: "080336006756",
      school_name: "George Washington",
      phone: '303.399.4532',
      address: '1234 S Monaco St.',
      distance: 5.8,
      low_Grade: '9',
      high_Grade: '12',
      level: 'High',
      is_Charter: false,
      is_Magnet: false,
      is_Private: false
    }];

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps.schools).toEqual(expectedSchools)
  });
});

describe('mapDispatchToProps', () => {

  it('should dispatch addSchools when called', () => {
    const mockSchools = {
      school_id: "080336006756",
      key: "080336006756",
      school_name: "George Washington",
      phone: '303.399.4532',
      address: '1234 S Monaco St.',
      distance: 5.8,
      low_Grade: '9',
      high_Grade: '12',
      level: 'High',
      is_Charter: false,
      is_Magnet: false,
      is_Private: false
    }
    const mockDispatch = jest.fn();
    const actionToDispatch = addSchools(mockSchools);
  
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addSchools({
      school_id: "080336006756",
      key: "080336006756",
      school_name: "George Washington",
      phone: '303.399.4532',
      address: '1234 S Monaco St.',
      distance: 5.8,
      low_Grade: '9',
      high_Grade: '12',
      level: 'High',
      is_Charter: false,
      is_Magnet: false,
      is_Private: false
    });
  
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  });


  it('should get schools from api when fetchAllSchools is called', async () => {
    const wrapper = shallow(<SearchForm/>);
    const mockSchools = [{
      school_id: "080336006756",
      key: "080336006756",
      school_name: "George Washington",
      phone: '303.399.4532',
      address: '1234 S Monaco St.',
      distance: 5.8,
      low_Grade: '9',
      high_Grade: '12',
      level: 'High',
      is_Charter: false,
      is_Magnet: false,
      is_Private: false
    }];
    const mockState = {"errorMsg": "", "latLocation": "", "level": "", "locState": "", "longLocation": "", "maxDistance": "", "render": false}
    const mockQuery = {   locState: 'CO',
    level: 'High',
    latLocation: '34.7838456',
    longLocation: '-104.7864334',
    maxDistance: '20'
  }
    const expectedSchools = [{
      school_id: "080336006756",
      key: "080336006756",
      school_name: "George Washington",
      phone: '303.399.4532',
      address: '1234 S Monaco St.',
      distance: 5.8,
      low_Grade: '9',
      high_Grade: '12',
      level: 'High',
      is_Charter: false,
      is_Magnet: false,
      is_Private: false
    }];

    expect(wrapper.state()).toEqual(mockState);

    await wrapper.instance().searchSchools(mockQuery);

    expect(wrapper.state()).toEqual(mockState)
  })
});