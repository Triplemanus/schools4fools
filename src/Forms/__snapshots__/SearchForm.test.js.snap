// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot with all data passed in correctly 1`] = `
<header>
  <h1>
    Search Form
  </h1>
  <form
    id="SearchForm"
  >
    <input
      name="locState"
      onChange={[Function]}
      placeholder="State"
      type="text"
      value=""
    />
    <input
      name="level"
      onChange={[Function]}
      placeholder="Level"
      value=""
    />
    <input
      name="latLocation"
      onChange={[Function]}
      placeholder="Latitude"
      value=""
    />
    <input
      name="longLocation"
      onChange={[Function]}
      placeholder="Longitude"
      value=""
    />
    <input
      name="maxDistance"
      onChange={[Function]}
      placeholder="Distance"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Search Schools
    </button>
  </form>
</header>
`;
